src/sentry/api/bases/chunk.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import absolute_import
+
+from sentry.models import File, FileBlob, FileBlobOwner
+from sentry.models.file import ChunkFileState
+
+
+class ChunkAssembleMixin(object):
+    def _create_file_response(self, state, missing_chunks=[]):
+        """
+        Helper function to create response for assemble endpoint
+        """
+        return {
+            'state': state,
+            'missingChunks': missing_chunks
+        }
+
+    def _check_chunk_ownership(self, organization, file_blobs, chunks, file_exists):
+        # Check the ownership of these blobs with the org
+        all_owned_blobs = FileBlobOwner.objects.filter(
+            blob__in=file_blobs,
+            organization=organization
+        ).select_related('blob').all()
+
+        owned_blobs = []
+        for owned_blob in all_owned_blobs:
+            owned_blobs.append((owned_blob.blob.id, owned_blob.blob.checksum))
+
+        # If the request does not contain any chunks for a file
+        # we return nothing since this should never happen only
+        # if the client sends an invalid request
+        if len(chunks) == 0:
+            return self._create_file_response(
+                ChunkFileState.NOT_FOUND
+            )
+        # Only if this org already has the ownership of all blobs
+        # and the count of chunks is the same as in the request
+        # and the file already exists, we say this file is OK
+        elif len(file_blobs) == len(owned_blobs) == len(chunks) and file_exists:
+            return self._create_file_response(
+                ChunkFileState.OK
+            )
+        # If the length of owned and sent chunks is not the same
+        # we return all missing blobs
+        elif len(owned_blobs) != len(chunks):
+            # Create a missing chunks array which we return as response
+            # so the client knows which chunks to reupload
+            missing_chunks = set(chunks)
+            for blob in owned_blobs:
+                if blob[1] in missing_chunks:
+                    missing_chunks.discard(blob[1])
+            # If we have any missing chunks at all, return it to the client
+            # that we need them to assemble the file
+            if len(missing_chunks) > 0:
+                return self._create_file_response(
+                    ChunkFileState.NOT_FOUND,
+                    missing_chunks
+                )
+
+    def _check_file_blobs(self, organization, checksum, chunks):
+        files = File.objects.filter(
+            checksum=checksum
+        ).select_related('blobs').all()
```

Comment:
I think you’re looking for `prefetch_related` here. Otherwise, below you’re doing an `O(n)` for each file to fetch the blobs.